core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -1930,12 +1931,24 @@ protected SqlSelect createSourceSelectForUpdate(SqlUpdate call) {
     }
     SqlNode sourceTable = call.getTargetTable();
     SqlIdentifier alias = call.getAlias();
+    SqlNode from = call.getFrom();
     if (alias != null) {
       sourceTable =
           SqlValidatorUtil.addAlias(
               sourceTable,
               alias.getSimple());
     }
+    if (from != null) {
+      sourceTable = new SqlJoin(
+          SqlParserPos.ZERO,
+          sourceTable,
+          SqlLiteral.createBoolean(false, SqlParserPos.ZERO),
+          JoinType.CROSS.symbol(SqlParserPos.ZERO),
+          from,
+          JoinConditionType.NONE.symbol(SqlParserPos.ZERO),
+          null
+      );
+    }
```

Comment:
See the note above about doing this in our own internal repository. It seems to me like it would be possible to implement this functionality there and then we wouldn't have to modify the fork for this.